### PR TITLE
Implement DPOP

### PR DESCRIPTION
**Pull request by 1485840691**
# What does this PR do?

Fixes # (issue)
https://github.com/huggingface/trl/issues/800

## Before submitting
- [y ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ y] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ y] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ y] Did you make sure to update the documentation with your changes?
- [ y] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
Nice, thanks, can you just add a very short subsection in the paper index section of the doc?

**Comment by 1485840691-eng** [comment]
> Nice, thanks, can you just add a very short subsection in the paper index section of the doc?

You mean an introduction of the core algo, in Section 4 of the paper, like following:
{B8391A26-6735-44ED-BB64-3689822FB3F1}

**Comment by 1485840691-eng** [comment]
I have compared DPO and DPOP in our dataset (train: about 1000 examples, eval: about 800 examples) in a text generation task from a supervised fine tuning model from Mistral-7B. The feature of the training data set is that the lexical token similarity between chosen and rejected pairs is above  0.8, meaning that the chosen and rejected pair is similar. Evaluation set is a mix of similar and different pairs. Following is my result 

DPO
{
"predict_bleu-4": 18.6497219140625,
"predict_rouge-1": 37.52502453125,
"predict_rouge-2": 21.85992626953125,
"predict_rouge-l": 22.51841951171875,
}

DPOP
{
"predict_bleu-4": 49.240244394531246,
"predict_rouge-1": 51.01286296875,
"predict_rouge-2": 35.5063100390625,
"predict_rouge-l": 54.486008945312506,
}

**Comment by 1485840691** [comment]
> > Nice, thanks, can you just add a very short subsection in the paper index section of the doc?
> 
> You mean an introduction of the core algo, in Section 4 of the paper, like following: {B8391A26-6735-44ED-BB64-3689822FB3F1}

@qgallouedec  would like to clarify is this what you need?

**Comment by 1485840691** [comment]
> I have compared DPO and DPOP in our dataset (train: about 1000 examples, eval: about 800 examples) in a text generation task from a supervised fine tuning model from Mistral-7B. The feature of the training data set is that the lexical token similarity between chosen and rejected pairs is above 0.8, meaning that the chosen and rejected pair is similar. Evaluation set is a mix of similar and different pairs. Following is my result
> 
> DPO { "predict_bleu-4": 18.6497219140625, "predict_rouge-1": 37.52502453125, "predict_rouge-2": 21.85992626953125, "predict_rouge-l": 22.51841951171875, }
> 
> DPOP { "predict_bleu-4": 49.240244394531246, "predict_rouge-1": 51.01286296875, "predict_rouge-2": 35.5063100390625, "predict_rouge-l": 54.486008945312506, }

Correct: the # of training examples is 200K, # of evaluation examples is 90K. The beta parameter is 0.1.

**Review by albertvillanova — COMMENTED** [review]
Sorry for the very long silence here @1485840691 — this one slipped through our backlog, and your follow-up question was never answered. 

To answer it now: yes, just a short subsection under `## Direct Policy Optimization` in `docs/source/paper_index.md`, in the style of the DiscoPOP one — two or three sentences plus a `DPOConfig` snippet, no need to reproduce the paper's derivation. 

In the meantime the DPO loss has been fully refactored: `DPOTrainer.dpo_loss` no longer exists and losses are dispatched in a `for loss_type in self.loss_types` loop, so this branch conflicts and can't be rebased mechanically — the test here would also fail with an `AttributeError`. 

Two notes for a fresh PR: DPOP should be its own `loss_type="dpop"` branch rather than a global `dpop_penalty_coef`, which as written silently perturbs `ipo`/`hinge`/`apo_*`/`discopop` as well; and the test should assert the penalty actually bites (loss strictly larger when `chosen_logratios < 0`) instead of only checking `torch.isfinite`. Your math is right for the sigmoid case.

I'm closing this as stale rather than unwanted — a PR against current `main` would be very welcome.
